src/crimson/os/seastore/nvmedevice/nvmedevice.h

```diff
@@ -0,0 +1,329 @@
+ * We selectively adopted features among various NVMe features to ease
+ * implementation. And also, NVMeBlockDevice provides generic command submission
+ * APIs for IO and Admin commands. Please use pass_through_io() and pass_admin()
+ * to do it.
+ *
+ * For more information about NVMe protocol, refer https://nvmexpress.org/
+ */
+
+struct nvme_identify_command_t {
+  uint32_t common_dw[10];
+  uint32_t cns : 8;
+  uint32_t reserved : 8;
+  uint32_t cntroller_id : 16;
+
+  static const uint8_t CNS_NAMESPACE = 0x00;
+  static const uint8_t CNS_CONTROLLER = 0x01;
+};
+
+struct nvme_admin_command_t {
+  union
+  {
+    nvme_passthru_cmd common_cmd;
+    nvme_identify_command_t identify_cmd;
+  };
+
+  static const uint8_t OPCODE_IDENTIFY = 0x06;
+};
+
+struct nvme_version_t {
+  uint32_t major_ver : 16;
+  uint32_t minor_ver : 8;
+  uint32_t tertiary_ver : 8;
+};
+
+struct admin_command_support_t {
+  uint16_t unused : 5;
+  uint16_t support_directives : 1;
+  uint16_t unused2 : 10;
+};
+
+struct identify_controller_data_t {
+  union
+  {
+    struct
+    {
+      uint8_t raw[1024];
+    };
+    struct
+    {
+      uint8_t unused[80];
+      nvme_version_t version;
+      uint8_t unused2[172];
+      admin_command_support_t oacs;
+    };
+  };
+};
+
+struct identify_namespace_data_t {
+  union
+  {
+    struct
+    {
+      uint8_t raw[4096];
+    };
+    struct
+    {
+      uint8_t unused[64];
+      uint16_t npwg;
+      uint16_t npwa;
+    };
+  };
+};
+
+using NVMePassThroughCommand = nvme_passthru_cmd;
+
+using read_ertr = crimson::errorator<
+  crimson::ct_error::input_output_error,
+  crimson::ct_error::invarg,
+  crimson::ct_error::enoent,
+  crimson::ct_error::erange>;
+
+using write_ertr = crimson::errorator<
+  crimson::ct_error::input_output_error,
+  crimson::ct_error::invarg,
+  crimson::ct_error::ebadf,
+  crimson::ct_error::enospc>;
+
+using open_ertr = crimson::errorator<
+  crimson::ct_error::input_output_error,
+  crimson::ct_error::invarg,
+  crimson::ct_error::enoent>;
+
+using nvme_command_ertr = crimson::errorator<
+  crimson::ct_error::input_output_error>;
+
+struct io_context_t {
+  iocb cb;
+  bool done = false;
+};
+
+/*
+ * Interface between NVMe SSD and its user.
+ *
+ * NVMeBlockDevice provides not only the basic APIs for IO, but also helper APIs
+ * to accelerate SSD IO performance and reduce system overhead. By aggresively
+ * utilizing and abstract useful features of latest NVMe SSD, it helps user ease
+ * to get high performance of NVMe SSD and low system overhead.
+ *
+ * Various implementations with different interfaces such as POSIX APIs, Seastar,
+ * and SPDK, are available.
+ */
+class NVMeBlockDevice {
+protected:
+  uint64_t size = 0;
+  uint64_t block_size = 0;
+
+  uint64_t write_granularity = 4096;
+  uint64_t write_alignment = 4096;
+
+public:
+  NVMeBlockDevice() {}
+  virtual ~NVMeBlockDevice() = default;
+
+  template <typename T>
+  static NVMeBlockDevice *create() {
```

Comment:
unique_ptr.  I don't really want any bare pointers as part of interfaces.

Reply:
OK